internal/iam/repository_grant_scope.go

```diff
@@ -0,0 +1,385 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package iam
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/oplog"
+)
+
+// AddRoleGrantScopes will add role grant scopes associated with the role ID in
+// the repository. No options are currently supported. Zero is not a valid value
+// for the WithVersion option and will return an error.
+func (r *Repository) AddRoleGrantScopes(ctx context.Context, roleId string, roleVersion uint32, grantScopes []string, _ ...Option) ([]*RoleGrantScope, error) {
+	const op = "iam.(Repository).AddRoleGrantScopes"
+	if roleId == "" {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing role id")
+	}
+	if len(grantScopes) == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing grant scopes")
+	}
+	if roleVersion == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing version")
+	}
+
+	role := allocRole()
+	role.PublicId = roleId
+
+	scope, err := role.GetScope(ctx, r.reader)
```

Comment:
I think, these reads (before you've gotten a Ticket in the transaction) are a race condition.  From `GetScope` to line 68 need to move into the Tx after you've called GetTicket.

Reply:
We use the passed-in `roleVersion` so it should be fine.